engines/freescape/objects/geometricobject.cpp

```diff
@@ -0,0 +1,347 @@
+		break;
+	case DownPyramid:
+		boundingBox.expand(origin + Math::Vector3d(size.x(), size.y(), 0));
+		boundingBox.expand(origin + Math::Vector3d(0, size.y(), 0));
+		boundingBox.expand(origin + Math::Vector3d(0, size.y(), size.z()));
+		boundingBox.expand(origin + Math::Vector3d(size.x(), size.y(), size.z()));
+
+		boundingBox.expand(origin + Math::Vector3d((*ordinates)[2], 0, (*ordinates)[1]));
+		boundingBox.expand(origin + Math::Vector3d((*ordinates)[0], 0, (*ordinates)[1]));
+		boundingBox.expand(origin + Math::Vector3d((*ordinates)[0], 0, (*ordinates)[3]));
+		boundingBox.expand(origin + Math::Vector3d((*ordinates)[2], 0, (*ordinates)[3]));
+		break;
+	case NorthPyramid:
+		boundingBox.expand(origin + Math::Vector3d(0, size.y(), 0));
+		boundingBox.expand(origin + Math::Vector3d(size.x(), size.y(), 0));
+		boundingBox.expand(origin + Math::Vector3d(size.x(), 0, 0));
+
+		boundingBox.expand(origin + Math::Vector3d((*ordinates)[0], (*ordinates)[3], size.z()));
+		boundingBox.expand(origin + Math::Vector3d((*ordinates)[2], (*ordinates)[3], size.z()));
+		boundingBox.expand(origin + Math::Vector3d((*ordinates)[2], (*ordinates)[1], size.z()));
+		boundingBox.expand(origin + Math::Vector3d((*ordinates)[0], (*ordinates)[1], size.z()));
+		break;
+	case SouthPyramid:
+		boundingBox.expand(origin + Math::Vector3d(0, 0, size.z()));
+		boundingBox.expand(origin + Math::Vector3d(size.x(), 0, size.z()));
+		boundingBox.expand(origin + Math::Vector3d(size.x(), size.y(), size.z()));
+
+		boundingBox.expand(origin + Math::Vector3d(0, size.y(), size.z()));
+		boundingBox.expand(origin + Math::Vector3d((*ordinates)[0], (*ordinates)[1], 0));
+		boundingBox.expand(origin + Math::Vector3d((*ordinates)[2], (*ordinates)[1], 0));
+		boundingBox.expand(origin + Math::Vector3d((*ordinates)[2], (*ordinates)[3], 0));
+		boundingBox.expand(origin + Math::Vector3d((*ordinates)[0], (*ordinates)[3], 0));
+		break;
+	}
+}
+
+GeometricObject::~GeometricObject() {
+	delete conditionSource;
+	delete colours;
+	delete ordinates;
+}
+
+bool GeometricObject::isDrawable() { return true; }
+bool GeometricObject::isPlanar() {
+	Type t = this->getType();
+	return (t >= Object::Line) || t == Object::Rectangle || !size.x() || !size.y() || !size.z();
+}
+
+bool GeometricObject::collides(const Math::AABB &_boundingBox) {
+	if (isDestroyed() || isInvisible() || !boundingBox.isValid() || !_boundingBox.isValid())
+		return false;
+
+	return (boundingBox.getMax().x() > _boundingBox.getMin().x() &&
+			boundingBox.getMin().x() < _boundingBox.getMax().x() &&
+			boundingBox.getMax().y() > _boundingBox.getMin().y() &&
+			boundingBox.getMin().y() < _boundingBox.getMax().y() &&
+			boundingBox.getMax().z() > _boundingBox.getMin().z() &&
+			boundingBox.getMin().z() < _boundingBox.getMax().z());
+}
+
+void GeometricObject::draw(Freescape::Renderer *gfx) {
+	// debug("Drawing %d of type %d", this->getObjectID(), this->getType());
```

Comment:
maybe put this into runtime condition instead commented code

Reply:
I will remove this since it is too verbose to be useful.

Reply:
Done!